File: src/moai_adk/resources/templates/.moai/scripts/file_analyzer.py
#!/usr/bin/env python3
"""
File Analyzer Module for MoAI Commit Helper
Handles file change analysis and classification

@TASK:FILE-ANALYSIS-001
@FEATURE:FILE-CHANGE-ANALYSIS-001
@API:ANALYZE-FILES
@TRUST:READABLE
"""

from typing import Any


class FileAnalyzer:
    """File change analyzer for smart commit processing"""

    # Constants for file type classification
    PYTHON_EXTENSIONS = {".py", ".pyx", ".pyi"}
    DOCUMENTATION_EXTENSIONS = {".md", ".rst", ".txt", ".doc", ".docx"}
    CONFIG_EXTENSIONS = {".json", ".yaml", ".yml", ".toml", ".ini", ".cfg"}
    TEST_PATTERNS = {"test_", "_test", "tests/", "test/"}

    def classify_file_change(self, status: str) -> str:
        """Classify file change type based on git status"""
        status_map = {
            "A": "added",
            "M": "modified",
            "D": "deleted",
            "R": "renamed",
            "C": "copied",
        }

        # Get first character of status for classification
        first_char = status[0] if status else "?"
        return status_map.get(first_char, "unknown")

    def analyze_file_changes(self, files: list[dict[str, Any]]) -> dict[str, Any]:
        """Analyze file changes and provide detailed classification"""
        if not files:
            return {
                "summary": {"added": 0, "modified": 0, "deleted": 0, "renamed": 0},
                "categories": {"python": 0, "docs": 0, "config": 0, "tests": 0, "other": 0},
                "total_files": 0,
                "is_simple_change": True,
            }

        summary = self._summarize_changes(files)
        categories = self._categorize_files(files)

        return {
            "summary": summary,
            "categories": categories,
            "total_files": len(files),
            "is_simple_change": self._is_simple_change(summary, categories),
            "has_tests": categories["tests"] > 0,
            "is_documentation_only": categories["docs"] > 0 and categories["python"] == 0,
        }

    def get_file_category(self, filename: str) -> str:
        """Get category of a single file"""
        filename_lower = filename.lower()

        # Check for test files first
        if any(pattern in filename_lower for pattern in self.TEST_PATTERNS):
            return "tests"

        # Check by extension
        extension = self._get_file_extension(filename)

        if extension in self.PYTHON_EXTENSIONS:
            return "python"
        elif extension in self.DOCUMENTATION_EXTENSIONS:
            return "docs"
        elif extension in self.CONFIG_EXTENSIONS:
            return "config"
        else:
            return "other"

    def suggest_commit_type(self, analysis: dict[str, Any]) -> str:
        """Suggest commit type based on file analysis"""
        categories = analysis["categories"]
        summary = analysis["summary"]

        # Guard clause: No changes
        if analysis["total_files"] == 0:
            return "chore"

        # Test-related changes
        if categories["tests"] > 0 and categories["python"] == 0:
            return "test"

        # Documentation-only changes
        if analysis["is_documentation_only"]:
            return "docs"

        # Configuration changes
        if categories["config"] > 0 and categories["python"] == 0:
            return "config"

        # New features (additions with Python files)
        if summary["added"] > 0 and categories["python"] > 0:
            return "feat"

        # Bug fixes or modifications
        if summary["modified"] > 0:
            return "fix" if analysis["is_simple_change"] else "refactor"

        # Deletions
        if summary["deleted"] > 0:
            return "remove"

        return "chore"

    def _summarize_changes(self, files: list[dict[str, Any]]) -> dict[str, int]:
        """Summarize changes by type"""
        summary = {"added": 0, "modified": 0, "deleted": 0, "renamed": 0}

        for file in files:
            file_type = file.get("type", "unknown")
            if file_type in summary:
                summary[file_type] += 1

        return summary

    def _categorize_files(self, files: list[dict[str, Any]]) -> dict[str, int]:
        """Categorize files by purpose"""
        categories = {"python": 0, "docs": 0, "config": 0, "tests": 0, "other": 0}

        for file in files:
            filename = file.get("filename", "")
            category = self.get_file_category(filename)
            categories[category] += 1

        return categories

    def _is_simple_change(self, summary: dict[str, int], categories: dict[str, int]) -> bool:
        """Determine if this is a simple change"""
        total_files = sum(summary.values())

        # Simple if few files changed
        if total_files <= 3:
            return True

        # Simple if only one category affected
        non_zero_categories = sum(1 for count in categories.values() if count > 0)
        if non_zero_categories == 1:
            return True

        return False

    def _get_file_extension(self, filename: str) -> str:
        """Extract file extension safely"""
        if "." not in filename:
            return ""

        return "." + filename.split(".")[-1].lower()

File: moai-adk-ts/src/scripts/analyzers/file-analyzer.ts
/**
 * @FEATURE-FILE-ANALYZER-001: TypeScript FileAnalyzer 구현
 * @연결: @TASK:ANALYSIS-001 → @FEATURE:FILE-ANALYSIS-001 → @API:ANALYZE-FILES
 */

import type { FileChange } from '../commit-helper';

export interface FileAnalysis {
  total_files: number;
  categories: {
    python: number;
    typescript: number;
    docs: number;
    config: number;
    tests: number;
    other: number;
  };
  is_simple_change: boolean;
  has_tests: boolean;
  is_documentation_only: boolean;
  summary?: {
    modified: number;
    added: number;
    deleted: number;
    renamed: number;
  };
}

export class FileAnalyzer {
  classifyFileChange(status: string): string {
    const statusCode = status.trim().charAt(0);

    switch (statusCode) {
      case 'M':
        return 'modified';
      case 'A':
        return 'added';
      case 'D':
        return 'deleted';
      case 'R':
        return 'renamed';
      case '?':
        return 'untracked';
      default:
        return 'unknown';
    }
  }

  analyzeFileChanges(files: FileChange[]): FileAnalysis {
    const categories = {
      python: 0,
      typescript: 0,
      docs: 0,
      config: 0,
      tests: 0,
      other: 0,
    };

    const summary = {
      modified: 0,
      added: 0,
      deleted: 0,
      renamed: 0,
    };

    let hasTests = false;
    let isDocumentationOnly = true;

    for (const file of files) {
      // 파일 카테고리 분류
      const category = this.categorizeFile(file.filename);
      categories[category]++;

      // 변경 유형 분류
      const changeType = file.type as keyof typeof summary;
      if (changeType in summary) {
        summary[changeType]++;
      }

      // 테스트 파일 검사
      if (this.isTestFile(file.filename)) {
        hasTests = true;
      }

      // 문서 전용 변경 검사
      if (!this.isDocumentationFile(file.filename)) {
        isDocumentationOnly = false;
      }
    }

    // 단순 변경 판단: 3개 이하 파일, 단일 카테고리
    const totalFiles = files.length;
    const nonZeroCategories = Object.values(categories).filter(
      count => count > 0
    ).length;
    const isSimpleChange = totalFiles <= 3 && nonZeroCategories <= 2;

    return {
      total_files: totalFiles,
      categories,
      is_simple_change: isSimpleChange,
      has_tests: hasTests,
      is_documentation_only: isDocumentationOnly && totalFiles > 0,
      summary,
    };
  }

  suggestCommitType(analysis: FileAnalysis): string {
    if (analysis.is_documentation_only) {
      return 'docs';
    }

    if (
      analysis.categories.tests > 0 &&
      analysis.categories.tests === analysis.total_files
    ) {
      return 'test';
    }

    if (analysis.categories.config > 0) {
      return 'chore';
    }

    if (analysis.summary?.added && analysis.summary.added > 0) {
      return 'feat';
    }

    if (analysis.summary?.modified && analysis.summary.modified > 0) {
      return 'fix';
    }

    return 'chore';
  }

  private categorizeFile(filename: string): keyof FileAnalysis['categories'] {
    const ext = this.getFileExtension(filename);

    // 테스트 파일
    if (this.isTestFile(filename)) {
      return 'tests';
    }

    // 문서 파일
    if (this.isDocumentationFile(filename)) {
      return 'docs';
    }

    // 설정 파일
    if (this.isConfigFile(filename)) {
      return 'config';
    }

    // 언어별 분류
    switch (ext) {
      case '.py':
        return 'python';
      case '.ts':
      case '.tsx':
      case '.js':
      case '.jsx':
        return 'typescript';
      default:
        return 'other';
    }
  }

  private isTestFile(filename: string): boolean {
    const basename = filename.toLowerCase();
    return (
      basename.includes('test') ||
      basename.includes('spec') ||
      basename.includes('__tests__') ||
      filename.endsWith('.test.ts') ||
      filename.endsWith('.test.js') ||
      filename.endsWith('.spec.ts')
    );
  }

  private isDocumentationFile(filename: string): boolean {
    const ext = this.getFileExtension(filename);
    const basename = filename.toLowerCase();

    return (
      ext === '.md' ||
      ext === '.rst' ||
      ext === '.txt' ||
      basename.includes('readme') ||
      basename.includes('changelog') ||
      basename.includes('docs/')
    );
  }

  private isConfigFile(filename: string): boolean {
    const basename = filename.toLowerCase();
    const configFiles = [
      'package.json',
      'tsconfig.json',
      'jest.config',
      'eslint',
      'prettier',
      '.env',
      'docker',
      'makefile',
      'pyproject.toml',
      'setup.py',
      'requirements.txt',
    ];

    return configFiles.some(config => basename.includes(config));
  }

  private getFileExtension(filename: string): string {
    const lastDot = filename.lastIndexOf('.');
    return lastDot !== -1 ? filename.substring(lastDot) : '';
  }
}
